solarforecastarbiter/datamodel.py

```diff
@@ -934,6 +938,227 @@ def __post_init__(self):
+        Parameters for the selected cost function.
+    """  # NOQA
+    error_range: Tuple[float, float]
+    cost_function: str
+    cost_function_parameters: Union[TimeOfDayCost, DatetimeCost, ConstantCost]
+
+    def _special_field_processing(self, model_field, val):
+        # support passing "inf", "-inf" as strings via json/dict
+        if model_field.name == 'error_range':
+            return float(val)
+        else:  # pragma: no cover
+            return val
+
+    def __post_init__(self):
+        if self.cost_function == 'timeofday':
+            if not isinstance(self.cost_function_parameters, TimeOfDayCost):
+                raise TypeError(
+                    "'cost_function_parameters' must be of type TimeOfDayCost "
+                    "for 'timeofday' cost function.")
+        elif self.cost_function == 'datetime':
+            if not isinstance(self.cost_function_parameters, DatetimeCost):
+                raise TypeError(
+                    "'cost_function_parameters' must be of type DatetimeCost "
+                    "for 'datetime' cost function.")
+        elif self.cost_function == 'constant':
+            if not isinstance(self.cost_function_parameters, ConstantCost):
+                raise TypeError(
+                    "'cost_function_parameters' must be of type ConstantCost "
+                    "for 'constant' cost function.")
+        else:
+            raise ValueError(
+                "'cost_function' must be one of 'timeofday', 'datetime', or"
+                " 'constant'")
+
+
+@dataclass(frozen=True)
+class ErrorBandCost(BaseModel):
+    """Cost that varies based on the error value. For each error band,
+    one of the other cost functions is applied to the errors within the band.
+    If an error value does not fall within any band ranges, no cost is
+    calculated for that error.
+
+    Parameters
+    ----------
+    bands : tuple of :py:class:`solarforecastarbiter.datamodel.CostBand`
+       Specification of the error bands and associated cost functions.
+       Each error is restricted to a single band/cost function, so the
+       order in bands determines which band is applied in ascending
+       priority. For example, if ``bands[0].error_range = (0, 2)``
+       and ``bands[1].error_range == (1, 3)``, the cost function of
+       bands[0] is applied for all errors from [0, 2] and bands[1]
+       is applied for errors from (2, 3].
+    """
+    bands: Tuple[CostBand, ...]
+
+
+@dataclass(frozen=True)
+class Cost(BaseModel):
+    """Specify how cost metrics should be calculated.
+
+    Parameters
+    ----------
+    name : str
+        Identifier for these cost parameters
+    type : str
+       The type of cost parameters that are included in `parameters`. One of
+       'timeofday', 'datetime', 'constant', or 'errorband'.
```

Comment:
also seems unnecessary

Reply:
ah, you want this for `from_dict`

Reply:
Yes. I think specifying the type is more clear especially when reading the dict/json.